mirgecom/boundary.py

```diff
@@ -602,3 +689,628 @@ def temperature_bc(self, state_minus, **kwargs):
+            discr=discr, btag=btag, gas_model=gas_model, state_minus=state_minus)
+
+        grad_cv_plus = self.grad_cv_bc(state_minus=state_minus,
+                                       grad_cv_minus=grad_cv_minus,
+                                       normal=normal, **kwargs)
+        grad_t_plus = self.grad_temperature_bc(grad_t_minus, normal)
+
+        # Note that [Mengaldo_2014]_ uses F_v(Q_bc, dQ_bc) here and
+        # *not* the numerical viscous flux as advised by [Bassi_1997]_.
+        f_ext = viscous_flux(state=state_plus, grad_cv=grad_cv_plus,
+                             grad_t=grad_t_plus)
+
+        return f_ext@normal
+
+
+class SymmetryBoundary(PrescribedFluidBoundary):
+    r"""Boundary condition implementing symmetry/slip wall boundary.
+
+    a.k.a. Reflective inviscid wall boundary
+
+    This class implements an adiabatic reflective slip boundary given
+    by
+    $\mathbf{q^{+}} = [\rho^{-}, (\rho{E})^{-}, (\rho\vec{V})^{-}
+    - 2((\rho\vec{V})^{-}\cdot\hat{\mathbf{n}}) \hat{\mathbf{n}}]$
+    wherein the normal component of velocity at the wall is 0, and
+    tangential components are preserved. These perfectly reflecting
+    conditions are used by the forward-facing step case in
+    [Hesthaven_2008]_, Section 6.6, and correspond to the characteristic
+    boundary conditions described in detail in [Poinsot_1992]_.
+
+    .. automethod:: adiabatic_wall_state_for_advection
+    .. automethod:: adiabatic_wall_state_for_diffusion
+    .. automethod:: adiabatic_slip_grad_av
+    """
+
+    def __init__(self):
+        """Initialize the boundary condition object."""
+        PrescribedFluidBoundary.__init__(
+            self, boundary_state_func=self.adiabatic_wall_state_for_advection,
+            inviscid_flux_func=self.inviscid_wall_flux,
+            viscous_flux_func=self.viscous_wall_flux,
+            boundary_temperature_func=self.temperature_bc,
+            boundary_gradient_cv_func=self.grad_cv_bc
+        )
+
+    def adiabatic_wall_state_for_advection(self, discr, btag, gas_model,
+                                           state_minus, **kwargs):
+        """Return state with opposite normal momentum."""
+        actx = state_minus.array_context
+        nhat = thaw(discr.normal(btag), actx)
+
+        mom_plus = \
+            (state_minus.momentum_density
+             - 2*(np.dot(state_minus.momentum_density, nhat)*nhat))
+
+        cv_plus = make_conserved(
+            state_minus.dim, mass=state_minus.mass_density,
+            energy=state_minus.energy_density, momentum=mom_plus,
+            species_mass=state_minus.species_mass_density
+        )
+        return make_fluid_state(cv=cv_plus, gas_model=gas_model,
+                                temperature_seed=state_minus.temperature)
+
+    def adiabatic_wall_state_for_diffusion(self, discr, btag, gas_model,
+                                           state_minus, **kwargs):
+        """Return state with 0 velocities and energy(Twall)."""
```

Comment:
"""Return state with opposite normal momentum."""

Reply:
For the inviscid part of the boundary treatment we return the opposite normal momentum because we use a Riemann solver to to calculate the fluxes.   We set the BC directly for the viscous part because we just calculate the viscous flux using the BC state, not an average or Riemann solver.

Reply:
Fixed with https://github.com/illinois-ceesd/mirgecom/pull/705